### DISCUSSION TITLE
How to read protobuf text format data to generated prost code

### DISCUSSION THREAD

**Discussion by zyddnys** [tokio-rs/prost#485] [Q&A]
I am using tonic gRPC library which uses prost to generate the underlying protobuf rust structs. Now I have a protobuf text format file which I want to load its data to a generate strcut. Can I do this? Do I need to use some other libraries?

**Comment by olix0r** [comment]
@zyddnys tonic includes a utility (as does Prost): `tonic::include_proto!("packagename");`

Take a look at tonic's examples:
 https://github.com/hyperium/tonic/blob/8084f4ea26cccf9bd2d96d2a81eaea490aaf603b/examples/src/helloworld/server.rs#L6-L8

**Comment by zyddnys** [comment]
This imports generated protobuf code, but I want to load actual data into a generated protobuf struct.
Example:
Protobuf file (hello.proto):
```protobuf
syntax = "proto3";

package testproto;

message TestData {
    string hello = 1;
}

message TestFile {
    repeated TestData data = 1;
}
```
Actual data (hello_data.txt)
```
data {
  hello: "hello world 1"
}
data {
  hello: "hello world 2"
}
```
I want something to parse hello_data.txt
```rust
pub mod test_proto {
    tonic::include_proto!("testproto");
}
fn readfile() {
    let fileio = std::fs::File::open("hello_data.txt").unwrap();
    let test_file: test_proto::TestFile = prost::from_file(fileio).unwrap();
    assert_eq!(test_file.data.first().unwrap().hello, "hello world 1");
}
``` 
The above example can be parsed using [google.protobuf.text_format](https://googleapis.dev/python/protobuf/latest/google/protobuf/text_format.html) in Python, can I do this in rust?

**Reply by andrewhickman** [reply]
If you don't mind the performance cost of reflection, there is [`parse_text_format`](https://docs.rs/prost-reflect/latest/prost_reflect/struct.DynamicMessage.html#method.parse_text_format`)

**Reply by ra314** [reply]
For posterity (since I needed to look at other people's code to figure out how to do this) here is an [example](https://github.com/danburkert/snazzy/pull/3) of using DynamicMessage::parse_text_format.
